web/source/osk/visualKeyboard.ts

```diff
@@ -1836,7 +1836,9 @@ namespace com.keyman.osk {
      **/
     highlightKey(key: KeyElement, on: boolean) {
       // Do not change element class unless a key
-      if(!key || (key.className == '') || (key.className.indexOf('kmw-key-row') >= 0)) return;
+      if(!key || !key.className || key.className.indexOf('kmw-key-row') >= 0) {
```

Comment:
In what circumstances can `key.className` not be a thing? That seems weird.

Reply:
Now that you say that... interesting how it's only happening on Android devices.

![Screen Shot 2021-02-12 at 12 30 44 PM](https://user-images.githubusercontent.com/25213402/107733674-287aed00-6d2e-11eb-9dd0-f644397c6157.png)

Perhaps it's related to how Android handles popup keys partly on the Android/Java side?  iOS handles it 100% on the web side, but Android doesn't.  That distinction would certainly help explain why it's only happening on Android devices.

If so, there's a chance we might expect 'sticky' highlighting on keys after subkey display.  Seen anything like that, @darcywong00?

Also of interest - it's only happening, exclusively, on 14.0.209-beta?  No other versions, but across multiple devices?

Reply:
> Also of interest - it's only happening, exclusively, on 14.0.209-beta? No other versions, but across multiple devices?

Note it appears to be happening in langtechcameroon keyboard only: `file:///data/user/0/info.langtechcameroon.keyboard/app_data/keyboard.html` which may make the limited number of versions less relevant.

Reply:
... wait, so that'd be within KeymanEngine, but a different host app?

Reply:
@MattGyverLee is this your app? Any thoughts?

Reply:
Analytics are creepy. :) Yes, that's my app. I actually have this bug on video in my [Bluestacks] emulator, but I hadn't investigated it.
https://youtu.be/UMImOzYeJfg?t=784 (Timestamp 13:04-13:11)
Do I need to fix something on my keyboard or bump versions?

Reply:
Is this a result of applying the super-specified CSS classnames that you recommended?

Reply:
This wouldn't have anything to do with CSS. What we really need to understand is the circumstances in which `highlightKey` can be called in which it appears that the `key` parameter is an object but is not a HTML element.

Looking at `findNearestKey()`, I see one potential place this could be going wrong. While we check in one place that we don't inadvertently return a non-Element node (by testing `className`):

https://github.com/keymanapp/keyman/blob/bf4f50f2de37b8e26391fe7bae8f1b1a4c0c8230/web/source/osk/visualKeyboard.ts#L1433-L1437

In another place we don't make that test:

https://github.com/keymanapp/keyman/blob/bf4f50f2de37b8e26391fe7bae8f1b1a4c0c8230/web/source/osk/visualKeyboard.ts#L1466-L1468

I am not sure that this is actually ever returning a non-Element but it seems that rather than catching the error here, we really should be locating where the bad data is coming from?

Reply:
@MattGyverLee I don't suppose you can get us a repro of this issue? We are having trouble tracking it down.

Reply:
A video wasn't good enough? :P I don't get debug data out of KMEA, only the things that my app throws.

Reply:
Yeah, I can't repro that exactly. But, are you using the mouse to control? That could be a clue -- we've only been testing with touch.

Reply:
As a matter of fact, I was. That could be part of it.
Bluestacks HyperV Emulator on Windows, using a mouse, and an old beta version of Keyman. How many devices are reporting it?

Reply:
52 reports from 35 users. So mouse is not the issue. A variety of Android devices, so I think we should get you onto a newer build of Keyman in case the issue has been resolved by other means (these are all on build 14.0.209-beta).

But I still cannot repro this in Chrome mobile emulator at https://keymanweb.com/?tier=beta&version=14.0.209#aal-latn,Keyboard_sil_cameroon_qwerty (this is an imperfect simulation so that could be masking the problem)

Reply:
Any updates on this?

Reply:
I haven't had time to test and push a new release yet. It may be closer to the end of this month.

Reply:
> But I still cannot repro this in Chrome mobile emulator at https://keymanweb.com/?tier=beta&version=14.0.209#aal-latn,Keyboard_sil_cameroon_qwerty (this is an imperfect simulation so that could be masking the problem)

I think this may be because this is related to popups, which are implemented Java-side in the app as opposed to KeymanWeb.

Reply:
I've just re-tested with BlueStacks native emulator (Keyman + Cameroon QWERTY) and I cannot reproduce the problem with 14.0.252-beta, following the exact steps as shown in the video.

Note that one difference is the popup is now in the correct location above the base key, which indicates that there are indeed changes between the two versions -- the one in the video has the popup overlapping the base key.

So I am hopeful that this issue is already resolved.